### DISCUSSION TITLE
What if we want to save multiple SVGs?

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#168] [Q&A]
<strike>When we discussed this issue (https://github.com/py5coding/py5generator/issues/137) ...</strike>

UPDATE: I tried to save two files with `begin_record`/`end_record` in two different frames but it looks like I made some mistake ... so I started thinking it wouldn't work because I was missing the `exit_sketch()` (which in fact is only necessary when using `SVG` as a `size()` renderer argument). 
Everything works as it used to in Processing Python mode...

**Comment by hx2A** [comment]
I believe you can create a SVG (or PDF) sequence like this:

```python
import py5

def setup():
    py5.size(500, 500, py5.SVG, 'out_####.svg')


def draw():
    # draw stuff

    if py5.frame_count == 10:
        py5.exit_sketch()
```

That would save 10 SVG files. Can you give that a try?

I believe you can also use the record functionality with a filename like  `'out_####.svg'` if you wish to write multiple files but not exit.

**Reply by villares** [reply]
It is nice that your example does work, I just added the needed  `py5.run_sketch()`, but I miss the ability to interactively choose the frames to to save...

It is no big deal, but it is less flexible than the raster exports.

And then, with PDFs, now that you provided the  `.next_page()` method, one can add pages instead of saving multiple files... hmmm, but not sure I could pause recording...

PS: I should add a `size(..., ..., SVG, 'out_####.svg')` example like yours to my materials :)

**Comment by hx2A** [comment]
> but I miss the ability to interactively choose the frames to to save...

Can you code something for that using the record functionality? Only record when a certain flag has been set?

Another approach is to save everything but delete what you don't want.

**Reply by villares** [reply]
I must be confused. I tried this example and it worked without `exit_sketch()`
I wonder why the the other example I had didn't work. Maybe some other bug....

```python
save_now = False
fator_escala = 3

def setup():
    size(400, 400)

def draw():
    global save_now
    if save_now:
        begin_record(output)
        output.scale(fator_escala)

    # Desenho aqui
    background(200, 255, 255)
    rect(random(0, width - 100),
         random(0, height - 100), 100, 100)

    if save_now:
        end_record()
        save_now = False


def save_svg():
    global output, save_now
    output = create_graphics(int(width * fator_escala),
                             int(height * fator_escala),
                             SVG, '{}.svg'.format(frame_count))
    save_now = True

def key_pressed():
    if key == 's':
        save_svg()
```

**Reply by hx2A** [reply]
Yes, this works! Glad you figured it out!!

**Reply by villares** [reply]
So, to clarify, `exit_sketch()` is only needed when choosing `SVG` as a renderer argument in `size()` right?

I mixed things up. I have always teached this flag strategy...

**Reply by hx2A** [reply]
The flag strategy is a good approach here.

When you have `SVG` as the renderer argument in `size()`, no window appears. The only way to exit the Sketch is by calling `exit_sketch()` somewhere.

**Comment by villares** [comment]
This "accumulated frames saved example" works fine with PDF, but not with SVG, any insights, @hx2A ?

I found in some previous notes that this was only possible with PDF, I think I'm going to "write off" this as a SVG export limitation (and warn in my teaching materials).

```python
fator_escala = 10
salvar_svg = False

def setup():
    size(200, 200)  # 2000px x 2000px with scale factor 10
    
def draw():
    if salvar_svg:
        svg_output.scale(fator_escala)
    fill(255, 32)
    if is_mouse_pressed:    
        circle(mouse_x, mouse_y, width / 10)        
        
def key_pressed():
    global salvar_svg, file_name, svg_output
    if key == 'r':    # 'r' starts and stops recording
        if not salvar_svg:
            salvar_svg = True
            background(200) # screen only
            file_name = f'{frame_count}.svg' 
            svg_output = create_graphics(
                width * fator_escala, height * fator_escala, SVG, file_name)
            begin_record(svg_output)
        else:
            salvar_svg = False
            end_record()
```

**Reply by hx2A** [reply]
If this works as you want for the PDF renderer but not the SVG renderer, it seems it is a limitation of Processing's SVG renderer. The Processing SVG renderer is really just a pass-through to another SVG library (https://xmlgraphics.apache.org/batik/, I recall) and what it can and cannot do is somewhat limited.
